go/apps/mixnet/queue.go

```diff
@@ -0,0 +1,206 @@
+	error
+}
+
+// The SendQueue structure.
+type SendQueue struct {
+	batchSize, ct int
+
+	network string
+
+	nextAddr   map[uint64]string
+	nextConn   map[uint64]net.Conn
+	sendBuffer map[uint64]*list.List
+
+	enqueue chan Queueable
+	err     chan sendQueueError
+}
+
+// NewSendQueue creates a new SendQueue structure.
+func NewSendQueue(network string, batchSize int) (sq *SendQueue) {
+	sq = new(SendQueue)
+	sq.batchSize = batchSize
+	sq.network = network
+
+	sq.nextAddr = make(map[uint64]string)
+	sq.nextConn = make(map[uint64]net.Conn)
+	sq.sendBuffer = make(map[uint64]*list.List)
+
+	sq.enqueue = make(chan Queueable)
+	sq.err = make(chan sendQueueError)
+	return sq
+}
+
+// Enqueue adds a queueable object, such as a message or directive, to the
+// send queue.
+func (sq *SendQueue) Enqueue(q *Queueable) {
+	sq.enqueue <- *q
+}
+
+// DoSendQueue atomically adds messages to a queue and transmits messages in
+// batches. Typically a message is a cell, but when the calling router is an
+// exit point, the message length is arbitrary. A batch is transmitted when
+// there are messages on batchSize distinct sender channels.
+func (sq *SendQueue) DoSendQueue(kill <-chan bool) {
+	for {
+		select {
+		case <-kill:
+			for _, c := range sq.nextConn {
+				c.Close()
+			}
+			return
+
+		case q := <-sq.enqueue:
+
+			// Set the next-hop address. We don't allow the destination address
+			// or conneciton to be over written in order to avoid accumulating
+			// stale connections on routers.
+			if _, def := sq.nextAddr[*q.Id]; !def && q.Addr != nil {
+				sq.nextAddr[*q.Id] = *q.Addr
+			}
+
+			if q.Dir != nil {
+				sq.err <- sendQueueError{*q.Id,
+					errors.New("directives not implemented")}
+
+			} else if _, def := sq.nextAddr[*q.Id]; !def && q.Msg != nil {
+				sq.err <- sendQueueError{*q.Id,
+					errors.New("request to send message without a destination")}
+
+			} else {
+
+				// Create a send buffer for the sender ID if it doesn't exist.
+				if _, def := sq.sendBuffer[*q.Id]; !def {
+					sq.sendBuffer[*q.Id] = list.New()
+				}
+				buf := sq.sendBuffer[*q.Id]
+
+				// The buffer was empty but now has a message ready; increment
+				// the counter.
+				if buf.Len() == 0 {
+					sq.ct++
+				}
+
+				// Add message to send buffer.
+				buf.PushBack(q.Msg)
+			}
+
+			// Transmit the message batch if it is full.
+			if sq.ct == sq.batchSize {
+				sq.dequeue()
+			}
+		}
+	}
+}
+
+// DoSendQueueErrorHandler handles errors produced by DoSendQueue. When this
+// is fully fleshed out, it will enqueue into the response queue a Directive
+// containing an error message. For now, just print out the error.
+func (sq *SendQueue) DoSendQueueErrorHandler(kill <-chan bool) {
+	for {
+		select {
+		case <-kill:
+			return
+		case err := <-sq.err:
+			fmt.Fprintf(os.Stderr, "send queue (%d): %s\n", err.id, err.Error())
+		}
+	}
+}
+
+// Send one message from each send buffer for each serial ID in a random
```

Comment:
Please start with the name of the function for consistency with exported methods. So, "dequeue sends one message..."